VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/PrivateSetAutoPropertyCanBeReadOnlyAutoProperty/PrivateSetAutoPropertyCanBeReadOnlyAutoPropertyAnalyzer.cs

```diff
@@ -0,0 +1,78 @@
+﻿using System;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using System.Collections.Generic;
+using System.Linq;
+
+
+namespace VSDiagnostics.Diagnostics.General.PrivateSetAutoPropertyCanBeReadOnlyAutoProperty
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class PrivateSetAutoPropertyCanBeReadOnlyAutoPropertyAnalyzer : DiagnosticAnalyzer
+    {
+        public const string DiagnosticId = nameof(PrivateSetAutoPropertyCanBeReadOnlyAutoPropertyAnalyzer);
+        internal const string Title = "Private Set AutoProperty is only set once. It can be made readonly.";
+        internal const string Message = "Private Set can be made ReadOnly";
+        internal const string Category = "General";
+        internal const DiagnosticSeverity Severity = DiagnosticSeverity.Warning;
+        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticId, Title, Message, Category, Severity, true);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        private List<SyntaxToken> _tokens = new List<SyntaxToken>();
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeSymbol, SyntaxKind.PropertyDeclaration);
+        }
+
+        private void AnalyzeSymbol(SyntaxNodeAnalysisContext context)
+        {
+            Diagnostic diagnostic = null;
+
+            var asProperty = context.Node as PropertyDeclarationSyntax;
+            if (asProperty != null)
+            {
+                diagnostic = HandleProperty(asProperty);
+            }
+
+            if (diagnostic != null)
+            {
+                context.ReportDiagnostic(diagnostic);
+            }
+        }
+
+        private Diagnostic HandleProperty(PropertyDeclarationSyntax propertyDeclaration)
+        {
+    
+            var setter = propertyDeclaration.AccessorList.Accessors.FirstOrDefault(s => s.IsKind(SyntaxKind.SetAccessorDeclaration) 
+                                                                                    && s.Modifiers.Any(m => m.IsKind(SyntaxKind.PrivateKeyword)));
+            if (setter != null)
+            {
+                if ( !TreeHasMethodsThatSetProperty(propertyDeclaration) )
+                {
+                    return Diagnostic.Create(Rule, setter.GetLocation(), "Property Setter", propertyDeclaration.Identifier);
```

Comment:
`Diagnostic.Create()` just creates a diagnostic (yes, really ;) )and takes 3 arguments: the first on is the `DiagnosticRule` we defined above which contains message, severity, ID, etc. The second argument is the location in the document where we want to show our warning (the squiggly lines you see in the IDE). The third argument is of type `params object[]` and contains the data you want to format your message with. If you'll look at other analyzers that use this third argument, you'll see that their message looks like this: `const string Message = "{0} blaat {1}`. It's the same idea as `string.Format()`: you replace the placeholders there with the data you pass in as arguments.